File: src/main/java/com/hazza/LeetCodeByJava/SegmentThree/No113PathSumII.java
package com.hazza.LeetCodeByJava.SegmentThree;

import com.hazza.LeetCodeByJava.datastructure.TreeNode;

import java.util.ArrayList;
import java.util.List;

/**
 * Created with IntelliJ IDEA.
 * Description: https://leetcode.com/problems/path-sum-ii/description/
 * User: HazzaCheng
 * Contact: hazzacheng@gmail.com
 * Date: 18-1-25
 * Time: 2:35 PM
 */
public class No113PathSumII {

    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        List<List<Integer>> res = new ArrayList<>();
        if (root == null) return res;
        treeDFS(root, sum, res, new ArrayList<>());

        return res;
    }

    private void treeDFS(TreeNode node, int sum, List<List<Integer>> res, List<Integer> path) {
        if (node == null) return;
        sum -= node.val;
        path.add(node.val);
        if (node.left == null && node.right == null && sum == 0) {
            res.add(new ArrayList<>(path));
            path.remove(path.size() - 1);
            return;
        }
        treeDFS(node.left, sum, res, path);
        treeDFS(node.right, sum, res, path);
        path.remove(path.size() - 1);
    }
}

File: src/main/python/leetcode_by_python/tree/No113_Path_Sum_II.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019-10-22
from typing import List

from src.main.python.leetcode_by_python.data_structure.TreeNode import TreeNode


class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        res = []
        self.dfs(root, sum, [], res)

        return res

    def dfs(self, node, sum, path, res):
        if not node:
            return

        sum -= node.val
        if not node.left and not node.right and sum == 0:
            res.append(path + [node.val])
            return
        self.dfs(node.left, sum, path + [node.val], res)
        self.dfs(node.right, sum, path + [node.val], res)
